[CORL-235] Allow a user to download their comment history

nick-funk opened the pull request:
## What does this PR do?

Allows a user to request to download their comments. This will trigger an email that they can then use to hit a landing page to download a CSV of their comment history.

## How do I test this PR?

- Have email development setup via Inbucket or similar
- Click on My Profile > Settings in the comment stream tabs
- Select "Request comment history"
- Check your email service for the email sent out
- Click the link in the email
- You should appear on a landing page to download your comments
- Select "Download My Comment History"
- An archive of your comments should download
- Unzip the archive and verify the comment data for accuracy

nick-funk committed:
Create preliminary UI for requesting a user's comment history

CORL-235

nick-funk committed:
Create preliminary download handler

CORL-235

nick-funk committed:
Create preliminary CSV exporter for downloading user's comments

CORL-235

nick-funk committed:
Create preliminary download comments flow with email and landing page

User can download their comments after requesting via an email.
Does not include limiters preventing users from spamming the download flow.

CORL-235

nick-funk committed:
Style the download comments landing page to match the spec

CORL-235

nick-funk committed:
Update the profile download comments to respect 14 day download limit

CORL-235

nick-funk committed:
Hook up request limiter on the comment download end points.

CORL-235

nick-funk committed:
Create sorry dialogue for when the download comments link is invalid/expired

CORL-235

nick-funk committed:
Create loading spinner for comment download landing page

CORL-235

nick-funk committed:
Create an optimistic updater for requesting a comment download

CORL-235

nick-funk committed:
Add localization for download comments landing page's sorry text

CORL-235

nick-funk committed:
Adjust downloaded comment format and contents

- Format the date to match tenant local
- Format dates to be using numeric values with date and time
- Add comment URL to exported CSV record

CORL-235

nick-funk committed:
Rename `createCSV` to `createExportedContent` in the download handler

CORL-235

nick-funk committed:
Make comment download callout full width and left justified

CORL-235

nick-funk committed:
Update stream snapshots to account for the download comments changes

CORL-235

nick-funk committed:
Merge remote-tracking branch 'origin/next' into feature/CORL-235

# Conflicts:
#	src/core/client/account/test/__snapshots__/resetPassword.spec.tsx.snap
#	src/core/client/stream/tabs/Profile/Settings/SettingsContainer.tsx
#	src/locales/en-US/stream.ftl

nick-funk committed:
Update snapshots to correspond to download comments changes

CORL-235

nick-funk committed:
Merge remote-tracking branch 'origin/next' into feature/CORL-235

nick-funk committed:
Add date filter to the comment download

Will filter by date of the issued token to only include comments before
that time.

CORL-235

nick-funk committed:
Clean up the layout of the list items in the download comments landing page

CORL-235

wyattjoh commented on src/core/client/account/routes/download/Download/DownloadRoute.tsx:
```diff
@@ -0,0 +1,56 @@
+  }
+  if (state !== "VALID" || error) {
+    return (
+      <>
```
If both the `<DownloadDescription />` and `<Sorry />` are wrapped in a `<HorizontalGutter size="double" />`, then why use `<></>`?

nick-funk replied:
Fixed up, pulled out redundant HorizontalGutter

wyattjoh commented on src/core/client/account/routes/download/Download/Loading.tsx:
```diff
@@ -0,0 +1,15 @@
+const Loading: FunctionComponent = () => (
+  <HorizontalGutter size="double">
+    <Flex justifyContent="center">
+      <Delay>
```
See the `<Loading />` component: `src/core/client/account/components/Loading.tsx`

nick-funk replied:
Used common Loading component

wyattjoh commented on src/core/client/account/routes/email/Confirm/ConfirmRoute.css:
```diff
@@ -0,0 +1,16 @@
+.root {
+  display: inline-block;
+
+  max-width: calc(70 * var(--mini-unit));
```
You modified the max-width in every route because you modified the entire account max-width to support the download page, why not also apply a max-width directly to the download page as well and don't constrain the width on the account?

wyattjoh committed:
fix: localization fixes

wyattjoh committed:
fix: formatting

wyattjoh committed:
fix: download file enhancements

wyattjoh committed:
fix: addressed download request loophole

nick-funk committed:
Update snapshots to address button changes on download comments landing page

CORL-235

nick-funk committed:
Simplify download comment landing page components

CORL-235

nick-funk committed:
Remove max-width calc's on account main layout and set them in child views

Sets the max-width settings on the download route since it's custom
set on all the othere routes (email and password).

CORL-235

wyattjoh commented on package-lock.json:
```diff
@@ -23831,7 +23831,7 @@
         "async": {
           "version": "1.5.2",
-          "resolved": "https://registry.npmjs.org/async/-/async-1.5.2.tgz",
+          "resolved": "http://registry.npmjs.org/async/-/async-1.5.2.tgz",
```
What version of `npm` are you running? Can you update `package.json:engines.npm` to match the most recent version?

nick-funk replied:
Running npm version 6.9.0.

I'll update it for sure.

nick-funk committed:
Bump npm version in package.json

CORL-235
